include/complex_quda.h

```diff
@@ -441,119 +433,74 @@ struct complex
       return *this;
     }
 
-  __host__ __device__ inline ValueType real() const volatile;
-  __host__ __device__ inline ValueType imag() const volatile;
   __host__ __device__ inline ValueType real() const;
   __host__ __device__ inline ValueType imag() const;
-  __host__ __device__ inline void real(ValueType) volatile;
-  __host__ __device__ inline void imag(ValueType) volatile;
   __host__ __device__ inline void real(ValueType);
   __host__ __device__ inline void imag(ValueType);
 };
 
-// TODO make cuFloatComplex and cuDoubleComplex protected
-// TODO see if returning references is a perf hazard
-
 template<>
-  struct complex <float> : public float2 //cuFloatComplex
+  struct complex <float> : public float2
 {
 public:
   typedef float value_type;
-  __host__ __device__ inline complex<float>() {};
-  __host__ __device__
-    inline complex<float>(const float & re, const float& im = float())
-    {
-      real(re);
-      imag(im);
-    }
+  complex<float>() = default;
+  constexpr complex<float>(const float & re, const float& im = float()) :
+    float2{re, im} { }
 
-  // For some reason having the following constructor
-  // explicitly makes things faster with at least g++
-  __host__ __device__
-    complex<float>(const complex<float> & z)
-    : float2(z){}
-
-  __host__ __device__
-    complex<float>(float2 z)
-    : float2(z){}
-  
-  template <class X>
-    inline complex<float>(const std::complex<X> & z)
-    {
-      real(z.real());
-      imag(z.imag());
-    }  
-
-  // Member operators
-  template <typename T>
-    __host__ __device__
-    inline volatile complex<float>& operator=(const complex<T> z) volatile
-    {
-      real(z.real());
-      imag(z.imag());
-      return *this;
-    }
+  template <typename X>
+  constexpr complex<float>(const std::complex<X> & z) :
+    float2{static_cast<float>(z.real()), static_cast<float>(z.imag())} { }
 
   template <typename T>
     __host__ __device__
-    inline complex<float>& operator=(const complex<T> z)
+    inline complex<float>& operator=(const complex<T> &z)
     {
       real(z.real());
       imag(z.imag());
       return *this;
     }
 
   __host__ __device__
-    inline complex<float>& operator+=(const complex<float> z)
+    inline complex<float>& operator+=(const complex<float> &z)
     {
       real(real()+z.real());
       imag(imag()+z.imag());
       return *this;
     }
 
   __host__ __device__
-    inline complex<float>& operator-=(const complex<float> z)
+    inline complex<float>& operator-=(const complex<float> &z)
```

Comment:
Just wondering: Why not just pass by value here?

Reply:
https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#f16-for-in-parameters-pass-cheaply-copied-types-by-value-and-others-by-reference-to-const

Reply:
I was intentionally making the `quda::complex` class match what is defined in the `std::complex`

Reply:
Got it.